### PR TITLE
test: register a custom variable in a custom Lua file and source it with `extra_lua_path`

### PR DESCRIPTION
**Pull request by kayx23**
### Description

* Updated doc for _Custom Variable._
* Added example and test to register a custom variable by sourcing a custom Lua file.

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: example/register_custom_var.lua [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    ngx.log(ngx.EMERG, "my hook works in http")
+    old_http_init(...)
+end
+
+local old_stream_init = apisix.stream_init
+apisix.stream_init = function(...)
+    register_custom_var()
+    ngx.log(ngx.EMERG, "my hook works in stream")
+    old_stream_init(...)
+end
```
Comment (Revolyssup) [inline example/register_custom_var.lua:42]: To show that the variable was really registered, should we try to access and log this in another phase and then grep and assert in the tests?

Comment (kayx23) [reply example/register_custom_var.lua:42]: Do you mean something like this?

```lua
local old_http_access_phase = apisix.http_access_phase
apisix.http_access_phase = function(...)
    ngx.log(ngx.EMERG, ctx.a6_route_labels)
    old_http_access_phase(...)
end
```

My understanding is that the custom variable `a6_route_labels` would only has value if a route is configured with the prop `labels`. 

The `register_var("a6_route_labels")` example is kept consistent with this test case: https://github.com/apache/apisix/blob/89c1858662e34c76de01c84b665c79fcb2edaeb8/t/plugin/http-logger-log-format.t#L431-L441

Please advise.

Comment (Revolyssup) [reply example/register_custom_var.lua:42]: Yes part of the suggestion was to PUT that route (just like in the test you mentioned) and then send request to that route, after which to make assertion. But I don't have a strong opinion here. If others think that showing how it's registered is good enough then I think its fine.

**Review by shreemaan-abhishek — APPROVED** [review]
Since we already have a test:
https://github.com/apache/apisix/blob/89c1858662e34c76de01c84b665c79fcb2edaeb8/t/plugin/http-logger-log-format.t#L431-L441

That shows the given example function for registering vars is correct. LGTM.

**Comment by Revolyssup** [comment]
> Since we already have a test:
> 
> https://github.com/apache/apisix/blob/89c1858662e34c76de01c84b665c79fcb2edaeb8/t/plugin/http-logger-log-format.t#L431-L441
> 
> That shows the given example function for registering vars is correct. This PR LGTM.

So we should reference this test in the doc

File: docs/en/latest/plugin-develop.md [hunk 2]
```diff
@@ -488,25 +487,33 @@ curl -i -X GET "http://127.0.0.1:9090/v1/plugin/example-plugin/hello"
...
-```
+```lua
 local core = require "apisix.core"
 
-core.ctx.register_var("a6_labels_zone", function(ctx)
+core.ctx.register_var("a6_route_labels", function(ctx)
     local route = ctx.matched_route and ctx.matched_route.value
     if route and route.labels then
-        return route.labels.zone
+      return route.labels
```
Comment (leslie-tsang) [inline docs/en/latest/plugin-develop.md:500]:
```suggestion
        return route.labels
```
Please standardize indentation style. `4` space will be please

File: example/register_custom_var.lua [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+apisix.http_init = function(...)
+    register_custom_var()
+    ngx.log(ngx.EMERG, "my hook works in http")
+    old_http_init(...)
+end
+
+local old_stream_init = apisix.stream_init
+apisix.stream_init = function(...)
+    register_custom_var()
+    ngx.log(ngx.EMERG, "my hook works in stream")
```
Comment (leslie-tsang) [inline example/register_custom_var.lua:40]:
```suggestion
    ngx.log(ngx.EMERG, "my hook works in stream")
```
please use `core.log` instead.

File: example/register_custom_var.lua (hunk 1 above)
Comment (leslie-tsang) [reply example/register_custom_var.lua:42]: We've already add test cases for this feature, and it doesn't seem necessary to do it again in the demo.
BTW, "my hook works in http" doesn't mean it actually works, it's just a log.

**Review by leslie-tsang — CHANGES_REQUESTED** [review]

**Comment by kayx23** [comment]
>  it doesn't seem necessary to do it again in the demo

@leslie-tsang this provides a working example for this issue (and doc): https://github.com/apache/apisix/issues/9878

File: example/register_custom_var.lua (hunk 1 above)
Comment (kayx23) [reply example/register_custom_var.lua:42]: Yeah I understand this doesn't cover the entire use case. The initial motivation of this was users having difficulties understanding how to register a variable with a custom Lua file that gets sourced (i.e. `lua_module_hook`). Once it's sourced and APISIX starts successfully (aka here it doesn't error out), the rest is easy. I agree with you that the feature of sourcing external files is covered in test, however.

**Comment by kayx23** [comment]
I agree this is more of an example than test coverage for the feature. Maybe we'd just update the docs alone to provide more instructions. https://github.com/apache/apisix/pull/10086#discussion_r1305408692

**Pull request closed without merging**